### PR TITLE
Fix the oversized list handling in ClusterStateCache

### LINKED ISSUE #17204: AttributeCache cannot handle large list
#### Problem
when reviewing bufferReaderCallback and AttributeCache, finding that when there is large buffered list, AttributeCache cannot use PacketBufferHandle with kMaxSecureSduLengthBytes to store that large list unless we improve PacketBufferHandle with chain or use Platform::ScopedMemoryBuffer when AttributeCache find it is list with size

### PR DESCRIPTION
**Pull request by yunhanw-google**
#### Problem
* Fixes #17204
#### Change overview
Use Platform::ScopedMemoryBuffer's calloc to allocate the actual size for attribute data in ClusterStateCache, update it across.
Fix API in ScopedMemoryBuffer's Get function.
Add mSize in ScopedMemoryBufferBase

#### Testing
Add unit test to get and validate the attributes including over-sized list combined from multiple chunks when using ClusterStateCache

### CODE REVIEW

File: src/app/ClusterStateCache.cpp [hunk 1]
```diff
@@ -417,8 +405,9 @@ void ClusterStateCache::GetSortedFilters(std::vector<std::pair<DataVersionFilter
                 }
                 else
                 {
-                    System::PacketBufferTLVReader bufReader;
-                    bufReader.Init(attributeIter.second.Get<System::PacketBufferHandle>().Retain());
+                    TLV::TLVReader bufReader;
+                    bufReader.Init(attributeIter.second.Get<Platform::ScopedMemoryBuffer<uint8_t>>().Get(),
```
Comment (mrjerryjohns) [inline src/app/ClusterStateCache.cpp:409]: See my comment below, this could be computed dynamically, avoiding the need to store an extra arg inside the scoped buffer.

File: src/controller/tests/data_model/TestRead.cpp [hunk 2]
```diff
@@ -785,19 +1152,139 @@ void TestReadInteraction::TestReadSubscribeAttributeResponseWithCache(nlTestSuit
...
-        NL_TEST_ASSERT(apSuite, cache.GetVersion(Test::kMockEndpoint2, Test::MockClusterId(2), version2) == CHIP_NO_ERROR);
+        app::ConcreteClusterPath clusterPath2(Test::kMockEndpoint2, Test::MockClusterId(2));
+        NL_TEST_ASSERT(apSuite, cache.GetVersion(clusterPath2, version2) == CHIP_NO_ERROR);
         NL_TEST_ASSERT(apSuite, version2.HasValue() && (version2.Value() == 2));
         Optional<DataVersion> version3;
-        NL_TEST_ASSERT(apSuite, cache.GetVersion(Test::kMockEndpoint1, Test::MockClusterId(2), version3) == CHIP_NO_ERROR);
+        app::ConcreteClusterPath clusterPath3(Test::kMockEndpoint1, Test::MockClusterId(2));
+        NL_TEST_ASSERT(apSuite, cache.GetVersion(clusterPath3, version3) == CHIP_NO_ERROR);
         NL_TEST_ASSERT(apSuite, version3.HasValue() && (version3.Value() == 2));
+
```
Comment (mrjerryjohns) [inline src/controller/tests/data_model/TestRead.cpp:1166]: This test is meant to test out the various cases of data versioning corner cases with the cache, and isn't the right home for testing out large lists. Moreover, I don't really understand why we have to do so many reads of the cache of different paths to validate a large list?

I'm not sure why we wouldn't have just put the updated test in `TestClusterStateCache.cpp`, specifically, line 233: that tests out sending lists. We could easily just amp up the count there from 4 to some much larger number that exceeds an IPv6 MTU.

File: src/app/ClusterStateCache.cpp [hunk 3]
```diff
@@ -44,20 +42,15 @@ CHIP_ERROR ClusterStateCache::UpdateCache(const ConcreteDataAttributePath & aPat
 
     if (apData)
     {
-        handle = System::PacketBufferHandle::New(chip::app::kMaxSecureSduLengthBytes);
-
-        writer.Init(std::move(handle), false);
-
+        Platform::ScopedMemoryBuffer<uint8_t> backingBuffer;
+        uint32_t totalBufSize = apData->GetTotalLength();
```
Comment (bzbarsky-apple) [inline src/app/ClusterStateCache.cpp:46]: But that assumes that the reader is positioned at the start of its buffer and the element ends at the end of the buffer, at least if we are to avoid over-allocating.  This _really_  matters for read responses with multiple attribute values in them.

Could we not just compute the size of the element directly and allocate the buffer to the right size?

File: src/lib/support/ScopedBuffer.h [hunk 4]
```diff
@@ -142,11 +148,13 @@ class ScopedMemoryBuffer : public Impl::ScopedMemoryBufferBase<MemoryManagement>
     inline T * Get() { return static_cast<T *>(Base::Ptr()); }
     inline T & operator[](size_t index) { return Get()[index]; }
 
-    inline const T * Get() const { return static_cast<T *>(Base::Ptr()); }
+    inline const T * Get() const { return static_cast<const T *>(Base::Ptr()); }
```
Comment (bzbarsky-apple) [inline src/lib/support/ScopedBuffer.h:151]: Why this change?  The Base::Ptr is non-const.  we are casting it to T*, and then returning it as a `const T*`, which are all valid operations....

Doing a single cast that both changes the type and the const-ness not ideal.

File: src/app/ClusterStateCache.cpp (hunk 3 above)
Comment (yunhanw-google) [reply src/app/ClusterStateCache.cpp:46]: from the current code path, OnAttributeData takes concrete path, and then pass the data to this UpdateCache, so this tlvreader would have only one element inside, there is no chance for multiple attribute?

File: src/lib/support/ScopedBuffer.h [hunk 5]
```diff
@@ -57,7 +56,9 @@ class ScopedMemoryBufferBase
         if (this != &other)
         {
             mBuffer       = other.mBuffer;
+            mSize         = other.mSize;
```
[the remark this answers is not in the document]
Comment (yunhanw-google) [reply src/lib/support/ScopedBuffer.h:59]: agree. I would prefer defining a new struct that contains the encapsulated tuple of the scoped buffer + size.

File: src/lib/support/ScopedBuffer.h (hunk 4 above)
Comment (yunhanw-google) [reply src/lib/support/ScopedBuffer.h:151]: in Base, there are two functions, one is  void * Ptr() { return mBuffer; }
   another is const void * Ptr() const { return mBuffer; }

File: src/lib/support/ScopedBuffer.h (hunk 5 above)
[the remark this answers is not in the document]
Comment (yunhanw-google) [reply src/lib/support/ScopedBuffer.h]: just add a wrapper to have the scope buffer and size. thanks

File: src/controller/tests/data_model/TestRead.cpp (hunk 2 above)
Comment (yunhanw-google) [reply src/controller/tests/data_model/TestRead.cpp:1166]: just add test for 1.5k size in TestClusterStateCache.
In testRead, I am still expecting to add the E2E test for the read on oversized list from chunked report in cluster state cache, and corresponding version read.

File: src/app/ClusterStateCache.cpp (hunk 3 above)
Comment (yunhanw-google) [reply src/app/ClusterStateCache.cpp]: agree, in updated PR, I have created GetElementTLVSize utility function and got the actual element size via using tlvwriter and CopyElement and GetLengthWritten

File: src/app/ClusterStateCache.cpp (hunk 1 above)
Comment (yunhanw-google) [reply src/app/ClusterStateCache.cpp]: follow your previous suggestion and use the wrapper to have scoped buffer and mSize, thanks

File: src/app/ClusterStateCache.cpp [hunk 6]
```diff
@@ -24,12 +24,26 @@
...
+    TLV::TLVReader reader;
+    reader.Init(*apData);
+    size_t totalBufSize = reader.GetTotalLength();
+    backingBuffer.Calloc(totalBufSize);
+    VerifyOrReturnError(backingBuffer.Get() != nullptr, CHIP_ERROR_NO_MEMORY);
+    TLV::ScopedBufferTLVWriter writer(std::move(backingBuffer), totalBufSize);
+    ReturnErrorOnFailure(writer.CopyElement(TLV::AnonymousTag(), reader));
+    aSize = writer.GetLengthWritten();
+    ReturnErrorOnFailure(writer.Finalize(backingBuffer));
+    backingBuffer.Release();
```
[the remark this answers is not in the document]
Comment (yunhanw-google) [reply src/app/ClusterStateCache.cpp:39]: yes, removed thanks.

File: src/app/BufferedReadCallback.cpp [hunk 7]
```diff
@@ -48,7 +48,7 @@ void BufferedReadCallback::OnReportEnd()
 CHIP_ERROR BufferedReadCallback::GenerateListTLV(TLV::ScopedBufferTLVReader & aReader)
 {
     TLV::TLVType outerType;
-    Platform::ScopedMemoryBuffer<uint8_t> backingBuffer;
+    Platform::ScopedMemoryBufferWithSize<uint8_t> backingBuffer;
```
[the remark this answers is not in the document]
Comment (yunhanw-google) [reply src/app/BufferedReadCallback.cpp:51]: I have made TLV::ScopedBufferTLVWriter/Reader takes ScopedMemoryBufferWithSize instead of ScopedMemoryBuffer

File: src/lib/core/CHIPTLV.h [hunk 8]
```diff
@@ -845,6 +845,13 @@ class DLL_EXPORT TLVReader
      * @return the stored backing store.
      */
     TLVBackingStore * GetBackingStore() { return mBackingStore; }
+
+    /**
+     * Return the total number of bytes for the TLV data
+     * @return the total number of bytes for the TLV data
+     */
+    uint32_t GetTotalLength() const { return mMaxLen; }
```
[the remark this answers is not in the document]
Comment (yunhanw-google) [reply src/lib/core/CHIPTLV.h:853]: agree, done, thanks

File: src/app/BufferedReadCallback.cpp (hunk 7 above)
[the remark this answers is not in the document]
Comment (yunhanw-google) [reply src/app/BufferedReadCallback.cpp:51]: no need for this change, removed, thanks

File: src/lib/support/ScopedBuffer.h [hunk 9]
```diff
@@ -160,5 +160,57 @@ class ScopedMemoryBuffer : public Impl::ScopedMemoryBufferBase<MemoryManagement>
...
+            mSize       = other.mSize;
+            other.mSize = 0;
+        }
+        ScopedMemoryBuffer<T>::operator=(std::move(other));
+        return *this;
+    }
+
+    virtual ~ScopedMemoryBufferWithSize() { mSize = 0; }
+
+    inline size_t GetSize() const { return mSize; }
```
[the remark this answers is not in the document]
Comment (yunhanw-google) [reply src/lib/support/ScopedBuffer.h:189]: Done, thanks

[the remark this answers is not in the document]
Comment (yunhanw-google) [reply src/lib/support/ScopedBuffer.h:189]: it is the size in bytes, updated it with BufferByteSize(), thanks

**Review by bzbarsky-apple — APPROVED** [review]

File: src/lib/support/ScopedBuffer.h [hunk 10]
```diff
@@ -160,5 +160,68 @@ class ScopedMemoryBuffer : public Impl::ScopedMemoryBufferBase<MemoryManagement>
...
+        {
+            mSize = 0;
+        }
+        return buffer;
+    }
+
+    ScopedMemoryBufferWithSize & Calloc(size_t elementCount)
+    {
+        ScopedMemoryBuffer<T>::Calloc(elementCount);
+        if (ScopedMemoryBuffer<T>::Get() != nullptr)
```
[the remark this answers is not in the document]
Comment (yunhanw-google) [reply src/lib/support/ScopedBuffer.h:202]: Done, thanks

**Review by yufengwangca — APPROVED** [review]
